Kickstarter-iOS/ViewModels/AppDelegateViewModelTests.swift

```diff
@@ -561,34 +566,140 @@ final class AppDelegateViewModelTests: TestCase {
+      self.vm.inputs.notificationAuthorizationStatusReceived(MockNotificationAutorizationStatus.notDetermined)
 
-      self.registerUserNotificationSettings.assertValueCount(3)
+      self.getNotificationAuthorizationStatus.assertValueCount(2)
+      self.authorizeForRemoteNotifications.assertValueCount(1)
+      self.registerForRemoteNotifications.assertValueCount(1)
+      self.unregisterForRemoteNotifications.assertValueCount(0)
+
+      self.vm.inputs.notificationAuthorizationCompleted()
+
+      self.getNotificationAuthorizationStatus.assertValueCount(3)
+      self.authorizeForRemoteNotifications.assertValueCount(1)
+      self.registerForRemoteNotifications.assertValueCount(1)
+      self.unregisterForRemoteNotifications.assertValueCount(0)
+
+      self.vm.inputs.notificationAuthorizationStatusReceived(MockNotificationAutorizationStatus.authorized)
+
+      self.getNotificationAuthorizationStatus.assertValueCount(3)
+      self.authorizeForRemoteNotifications.assertValueCount(1)
+      self.registerForRemoteNotifications.assertValueCount(2)
       self.unregisterForRemoteNotifications.assertValueCount(0)
     }
 
     self.vm.inputs.userSessionEnded()
 
-    self.registerUserNotificationSettings.assertValueCount(3)
+    self.registerForRemoteNotifications.assertValueCount(2)
     self.unregisterForRemoteNotifications.assertValueCount(1)
   }
 
+  fileprivate func count(_ string: String) -> ([String]) -> Int {
+    return { array in
+      return array.reduce(0, { acc, val in val == string ? acc + 1 : acc })
+    }
+  }
+
+  func testTrackingPushAuthorizationOptIn() {
+    let client = MockTrackingClient()
+
+    withEnvironment(currentUser: .template, koala: Koala(client: client)) {
+      self.vm.inputs.userSessionStarted()
+
+      XCTAssertEqual(0, client.events |> count("Triggered Push Permissions Dialog"))
+      XCTAssertEqual(0, client.events |> count("Confirmed Push Opt-In"))
+      XCTAssertEqual(0, client.events |> count("Dismissed Push Opt-In"))
+
+      self.vm.inputs.applicationDidEnterBackground()
+      self.vm.inputs.applicationWillEnterForeground()
+
+      XCTAssertEqual(0, client.events |> count("Triggered Push Permissions Dialog"))
+      XCTAssertEqual(0, client.events |> count("Confirmed Push Opt-In"))
+      XCTAssertEqual(0, client.events |> count("Dismissed Push Opt-In"))
+
+      self.vm.inputs.notificationAuthorizationStatusReceived(MockNotificationAutorizationStatus.notDetermined)
+
+      XCTAssertEqual(1, client.events |> count("Triggered Push Permissions Dialog"))
+      XCTAssertEqual(0, client.events |> count("Confirmed Push Opt-In"))
+      XCTAssertEqual(0, client.events |> count("Dismissed Push Opt-In"))
+
+      self.vm.inputs.notificationAuthorizationCompleted()
+      self.vm.inputs.notificationAuthorizationStatusReceived(MockNotificationAutorizationStatus.authorized)
+
+      XCTAssertEqual(1, client.events |> count("Triggered Push Permissions Dialog"))
```

Comment:
Thanks for explaining @bormind, we chatted about this a bit on our call but yes as @theginbin and @stephencelis are saying we probably want to be more explicit here about what will be emitted and in which order. I hear what you're saying about the autonomy of the test case but that isn't really a problem. If a change was made later on and more tracking events were added to `userSessionStarted` we would want to know about them and we would want our test to fail.

In fact in such a situation, where more tracking was performed on the session starting, we could actually rename the the test from `testTrackingPushAutorizationOptOut` to `testUserSessionStartedTrackingEvents` for example and test all of the various emissions that we expect to happen at that time. For now this is not necessary.

Lastly, and @stephencelis @theginbin correct me if I'm wrong, but introducing a helper function into a test generally requires that we _also_ test that function (however simple it may be?) otherwise it _can_ introduce another element of uncertainty in the test? Or is this not really an issue?